File: leetcode/graph/269_Alien_Dictionary/codes/graph.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <queue>
using namespace std;

class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char, unordered_set<char>> graph;
        unordered_map<char, int> in_degree;

        for (string& word : words)
            for (char c : word)
                in_degree[c] = 0;

        for (int i = 0; i < words.size() - 1; ++i) {
            string& w1 = words[i], &w2 = words[i + 1];
            if (w1.size() > w2.size() && w1.find(w2) == 0)
                return "";
            for (int j = 0; j < min(w1.size(), w2.size()); ++j) {
                char c1 = w1[j], c2 = w2[j];
                if (c1 != c2) {
                    if (!graph[c1].count(c2)) {
                        graph[c1].insert(c2);
                        in_degree[c2]++;
                    }
                    break;
                }
            }
        }

        queue<char> q;
        for (auto& [c, deg] : in_degree)
            if (deg == 0) q.push(c);

        string res;
        while (!q.empty()) {
            char c = q.front(); q.pop();
            res += c;
            for (char nei : graph[c]) {
                if (--in_degree[nei] == 0)
                    q.push(nei);
            }
        }

        return res.size() == in_degree.size() ? res : "";
    }
};

File: leetcode/graph/269_Alien_Dictionary/codes/graph.py
from collections import defaultdict, deque
from typing import List

class Solution:
    def alienOrder(self, words: List[str]) -> str:
        graph = defaultdict(set)
        in_degree = {c: 0 for word in words for c in word}

        for i in range(len(words) - 1):
            w1, w2 = words[i], words[i + 1]
            if len(w1) > len(w2) and w1.startswith(w2):
                return ""
            for c1, c2 in zip(w1, w2):
                if c1 != c2:
                    if c2 not in graph[c1]:
                        graph[c1].add(c2)
                        in_degree[c2] += 1
                    break

        queue = deque([c for c in in_degree if in_degree[c] == 0])
        res = []

        while queue:
            c = queue.popleft()
            res.append(c)
            for nei in graph[c]:
                in_degree[nei] -= 1
                if in_degree[nei] == 0:
                    queue.append(nei)

        return "".join(res) if len(res) == len(in_degree) else ""
